python/pyarrow/_flight.pyx

```diff
@@ -289,6 +289,33 @@ cdef class Action(_Weakrefable):
                 type(action)))
         return (<Action> action).action
 
+    def serialize(self):
+        """Get the wire-format representation of this type.
+
+        Useful when interoperating with non-Flight systems (e.g. REST
+        services) that may want to return Flight types.
+
+        """
+        return GetResultValue(self.action.SerializeToString())
+
+    @classmethod
+    def deserialize(cls, serialized):
+        """Parse the wire-format representation of this type.
+
+        Useful when interoperating with non-Flight systems (e.g. REST
+        services) that may want to return Flight types.
+
+        """
+        cdef Action action = Action.__new__(Action)
+        action.action = GetResultValue(
+            CAction.Deserialize(tobytes(serialized)))
+        return action
+
+    def __eq__(self, Action other):
+        # TODO (qhoang): question for myself, do we need to push this down
```

Comment:
Yes, just inherit from `util::EqualityComparable` on the C++ side and implement `Equals` - then you can just declare an `operator==` in Cython to use instead of doing this

Or I guess you can do something like this: https://github.com/apache/arrow/blob/a5ecb0ff0774805b0f912e231eaedf42e7194c36/cpp/src/arrow/flight/types.h#L143-L150

Then just add the declaration in Cython (ActionType is actually missing its declaration): https://github.com/apache/arrow/blob/a5ecb0ff0774805b0f912e231eaedf42e7194c36/python/pyarrow/includes/libarrow_flight.pxd#L58

Reply:
Thank you so much, @lidavidm 
* Added Equals and operator== to C++ side.

Need your help on 3 queries:
1. For ActionType, currently, it is defined as a named-tuple, should we change it to be backed by CActionType? https://github.com/apache/arrow/blob/master/python/pyarrow/_flight.pyx#L293
2. Same with Criteria, it is not defined in _flight.pyx yet, should I add it in this MR as well?
3. And the last one, mostly for your hint: 
```python
# This one works
    result = flight.Result(b"result1")
    result2 = flight.Result.deserialize(result.serialize())
    assert result.body == result2.body

# But it does not for this one:
    assert result == result2

# Result::__eq__ is already defined in _flight.pyx. (I am getting myself familiar with unique_ptr in Cython)
```
Appreciate your help. I have been learning a lot in last week.

Reply:
1. Don't think so
2. If it's not in Python then don't add it
3. What exactly does not work, how does it fail, what have you tried? It seems the definition of Result is slightly off, it should dereference these before comparison, because right now it's comparing the pointers not the values: https://github.com/apache/arrow/blob/03e204c731186e133b41fdbfdfa896da5404a819/python/pyarrow/_flight.pyx#L377-L378

Reply:
Thank you so much, going to add notes for (1. and 2.) in the MR description.

That's exactly the key point. I was confused by unique_ptr syntax in Cython. Frankly, I had to go over a couple of Cython-101 documents specifically for this issue.

Thanks once again David.